asst2/CODE/my_malloc.c

```diff
@@ -25,137 +22,76 @@ void error_message(char* error, char* file, int line) {
 }
 
 void initialize_pages() {
 	for (int i = 0; i < num_pages; i++) {
 		pagedata* pdata = (pagedata*)myblock + i;	
-		write_free_page(pdata, i);
+		pg_write_pagedata(pdata, 0, NOT_OCC, NOT_OVF, i, 1);
 		metadata* mdata = (metadata*) ((uint8_t*)mem_space + i*page_size);
 		dm_write_metadata(mdata, num_segments, 0, 1);
 	}
 }
 
-void* find_free_page(size_t size, uint32_t curr_id) {
-	//find page for process...
-	size_t segments_alloc = size / SEGMENTSIZE + 1;
-	int free_page = -1;
-	int page_exist = -1;
-	int index = (segments_alloc > num_segments) ? num_pages : 0; // remove this
-	// check to allow longer allocations
-
-	//if exists, find first fit free space
-	for (; index < num_pages; index++) {
-		pagedata* pdata = (pagedata*)myblock + index;	
-		if ( !is_occupied_page(pdata) ) { //free page
-			if (free_page == -1) {
-				free_page = index;
-			}
-		}
-		else if (pdata->pid == curr_id) { //occupied page
-			if (page_exist == -1) {
-				page_exist = index; //for swapping later...
-			}
-			//traverse memory to find the first free block that can fit the size requested
-			metadata* start = (metadata*) (mem_space + index*page_size);
-			int seg_index = 0;
-			while ( seg_index < num_segments) {
-				metadata* curr = start + seg_index*SEGMENTSIZE;
-				uint8_t curr_seg = dm_block_size(curr);
-				if (!dm_block_occupied(curr) && (segments_alloc <= curr_seg ) ) {
-					dm_allocate_block(curr, segments_alloc);
-					//segment memory space = start + num_segments
-					//free segment = segment memory space + seg_index * SEGMENTSIZE;
-					return (void*) (curr + 1);
-				}
-				seg_index+=curr_seg;
-			}
-		}
-		//else not enough space, find any other pages allocated to pid
-	}
-
-	//we did not find page for current process
-	//so allocate page, then return pointer within page chunk
-	if ( (page_exist == -1 || curr_id == 0) && free_page != -1 ) {
-		write_occupied_page((pagedata*)myblock+free_page, curr_id, free_page);
-		metadata* curr = (metadata*) (mem_space + free_page*page_size);
-		uint8_t curr_seg = dm_block_size(curr);
-		if (segments_alloc <= curr_seg) { 
-			dm_allocate_block(curr, segments_alloc);
-			return (void*) (curr + 1);
-		}
-	}
-
-	return NULL;
-}
-
-//returns to the user a pointer to the amount of memory requested
 void* myallocate(size_t size, char* file, int line, int threadreq){
 	enter_scheduler(&timer_pause_dump);
+	
 	//if the user asks for 0 bytes, call error and return NULL
 	if (size <= 0){
 		error_message("Cannot malloc 0 or negative bytes", file, line);
 		return NULL;
 	}
+	
 	uint32_t curr_id = (threadreq == LIBRARYREQ) ? 0 : ( (tcb*) get_head(ready_q[curr_prio]) )->id;
 	if (firstMalloc == 1) { // first time using malloc
 		myblock = memalign(sysconf( _SC_PAGESIZE), MEMSIZE);
 		page_size = sysconf( _SC_PAGESIZE);
 		num_segments = page_size / SEGMENTSIZE;
-		num_pages = MEMSIZE / ( page_size + sizeof(pagedata));
+		num_pages = MEMSIZE / ( page_size + sizeof(pagedata) + sizeof(ht_entry) );
```

Comment:
I'm not sure this is right. We'll ultimately need to also keep track of pages on disk

Reply:
hmm i wouldn't consider those pages here tho right? bc we don't have that extra memory space yet?

Reply:
ok, that's fair. can you add a todo here so we don't forget?